pkg/sql/catalog/multiregion/region_config.go

```diff
@@ -468,3 +519,109 @@ func CanDropRegion(name catpb.RegionName, config RegionConfig) error {
+	for _, constraint := range constraints {
+		// TODO(janexing): We should also consider the case with contraints on zone
+		// level.
+		if constraint.Constraints[0].Key == "region" &&
+			constraint.Constraints[0].Value == homeRegion {
+			return constraint, true
+		}
+	}
+	return zonepb.ConstraintsConjunction{}, false
+}
+
+// getHomeRegionLeasePreference returns the lease preference and its index from
+// a list of lease preference for the given region name. If not found in the
+// list, it returns an empty lease preference and -1 as the index.
+func getHomeRegionLeasePreference(
+	leasePreferences []zonepb.LeasePreference, homeRegion string,
+) (zonepb.LeasePreference, int) {
+	for idx, leasePreference := range leasePreferences {
+		// TODO(janexing): We should also consider the case with contraints on zone
+		// level.
+		if leasePreference.Constraints[0].Value == homeRegion {
+			return leasePreference, idx
+		}
+	}
+	return zonepb.LeasePreference{}, -1
+}
+
+// validateZoneConfigExtension ensures that the zone config extensions don't
+// set configs that disagree with their home region.
+func validateZoneConfigExtension(
+	ext zonepb.ZoneConfig, zc zonepb.ZoneConfig, homeRegion string,
+) error {
+	if homeRegion == "" {
+		return nil
+	}
+
+	// If the extension does not inherit that field from the original zone config,
+	// we will use the one in the extension, and hence we need to validate that
+	// field.
+	// Lease preferences.
+	if !(ext.ShouldInheritLeasePreferences(&zc)) && len(zc.LeasePreferences) > 0 && len(ext.LeasePreferences) > 0 {
+		oriHomeRegionLeasePreference, idx1 := getHomeRegionLeasePreference(zc.LeasePreferences, homeRegion)
+		extHomeRegionLeasePreference, idx2 := getHomeRegionLeasePreference(ext.LeasePreferences, homeRegion)
+		if idx1 != -1 {
+			// If can' find a lease preference for the home region in the zone config
+			// extension's lease preferences.
+			if idx2 == -1 {
+				return errors.Newf("zone config extension cannot unset the home "+
+					"region (%v) lease preference: %v\n", homeRegion, oriHomeRegionLeasePreference)
+			}
+			if idx1 != idx2 {
+				return errors.Newf("zone config extension cannot change priority"+
+					" of lease preference for the home region: %v)\n", oriHomeRegionLeasePreference)
+			}
+			if oriHomeRegionLeasePreference.Constraints[0].Type != extHomeRegionLeasePreference.Constraints[0].Type {
+				return errors.Newf("zone config extension %v violates the home"+
+					" region (%v) lease preference: %v\n", extHomeRegionLeasePreference, homeRegion, oriHomeRegionLeasePreference)
+			}
+		}
+	}
+
+	// Replica constraints.
+	if !(ext.ShouldInheritConstraints(&zc)) && len(zc.Constraints) > 0 && len(ext.Constraints) > 0 {
+		oriHomeRegionReplicaConstraint, ok1 := getHomeRegionConstraintConjunction(zc.Constraints, homeRegion)
+		extHomeRegionReplicaConstraint, ok2 := getHomeRegionConstraintConjunction(ext.Constraints, homeRegion)
+		if ok1 {
```

Comment:
This is a little bit hard to read.

Can we do `if ok1 && !ok2 { ... }` and split out the logic below
Also probably better to do name `ok1` as `homeRegionConstraintExistsForOldZC` and `ok2` as `homeRegionConstraintExistsForExt` as something like, I prefer verbosity here to make it clear.

Reply:
I changed it to return the reference of the constraint, so the second return value is not very much needed.